lib/at_cmd_parser/src/at_cmd_decoder.c

```diff
@@ -0,0 +1,277 @@
+
+static void *at_CNUM_decode(struct at_param_list *param_list,
+			    u32_t valid_params);
+
+AT_CMD_DECODER_LIST_START(default_decoder_list)
+AT_CMD_DECODER_LIST_ENTRY("+CESQ", AT_CMD_CESQ_MODEL, at_CESQ_decode)
+AT_CMD_DECODER_LIST_ENTRY("%CESQ", AT_NOT_CESQ_MODEL, at_NOT_CESQ_decode)
+AT_CMD_DECODER_LIST_ENTRY("+CEREG:", AT_CMD_CEREG_MODEL, at_CEREG_decode)
+AT_CMD_DECODER_LIST_ENTRY("+CMT:", AT_CMD_CMT_MODEL, at_CMT_decode)
+AT_CMD_DECODER_LIST_ENTRY("+CNUM:", AT_CMD_CNUM_MODEL, at_CNUM_decode)
+AT_CMD_DECODER_LIST_END();
+
+static struct at_cmd_decoder_list const *at_cmd_decoders;
+static struct at_param_list             param_list;
+static at_cmd_decoder_handler_t         at_cmd_decoder_handler;
+
+static void *at_CMT_decode(struct at_param_list *param_list,
+			   u32_t valid_params)
+{
+	int err;
+
+	if (valid_params != 3) {
+		return NULL;
+	}
+
+	struct at_cmd_model_cmt *model =
+			k_malloc(sizeof(struct at_cmd_model_cmt));
+
+	if (!model) {
+		return NULL;
+	}
+	memset(model, 0, sizeof(struct at_cmd_model_cmt));
+
+
+	err = at_params_get_string(param_list, 0, model->sender_addr, 32);
+	err |= at_params_get_short(param_list, 1, &model->pdu_length);
+	err |= at_params_get_string(param_list, 2, model->pdu_data, 160);
+
+	return model;
+}
+
+static void *at_CESQ_decode(struct at_param_list *param_list,
+			    u32_t valid_params)
+{
+	int err;
+
+	if (valid_params != 6) {
+		return NULL;
+	}
+
+	struct at_cmd_model_cesq *model =
+			k_malloc(sizeof(struct at_cmd_model_cesq));
+
+	if (!model) {
+		return NULL;
+	}
+	memset(model, 0, sizeof(struct at_cmd_model_cesq));
+
+	err = at_params_get_short(param_list, 0, &model->rxlev);
+	err |= at_params_get_short(param_list, 1, &model->ber);
+	err |= at_params_get_short(param_list, 2, &model->rscp);
+	err |= at_params_get_short(param_list, 3, &model->ecno);
+	err |= at_params_get_short(param_list, 4, &model->rsrq);
+	err |= at_params_get_short(param_list, 5, &model->rsrp);
+
+	if (err) {
+		k_free(model);
+		return NULL;
+	} else {
+		return model;
+	}
+}
+
+static void *at_NOT_CESQ_decode(struct at_param_list *param_list,
+			       u32_t valid_params)
+{
+	int err;
+
+	if (valid_params != 2) {
+		return NULL;
+	}
+
+	struct at_not_model_cesq *model =
+			k_malloc(sizeof(struct at_not_model_cesq));
+
+	if (!model) {
+		return NULL;
+	}
+	memset(model, 0, sizeof(struct at_not_model_cesq));
+
+	err = at_params_get_short(param_list, 0, &model->rsrp);
+	err |= at_params_get_short(param_list, 1, &model->threshold_index);
+
+	if (err) {
+		k_free(model);
+		return NULL;
+	} else {
+		return model;
+	}
+}
+
+static void *at_CNUM_decode(struct at_param_list *param_list,
+			   u32_t valid_params)
+{
+	int err;
+
+	if (valid_params != 2) {
+		return NULL;
+	}
+
+	struct at_cmd_model_cnum *model =
+			k_malloc(sizeof(struct at_cmd_model_cnum));
+
+	if (!model) {
+		return NULL;
+	}
+	memset(model, 0, sizeof(struct at_cmd_model_cnum));
+
+	err = at_params_get_string(param_list, 0, model->numberx, 32);
```

Comment:
The memory is allocated in the dekoder, and then de-allocated in the at_cmd_decode function. As you see, all of the decoder functions are private. The idea is that the user shall call the decode function, when it has parsed and decoded the data it will call a callback of `type at_cmd_decoder_handler_t`, and when this callback returns the memory allocated for the model data is freed.